### PR TITLE
ODPR-649: Update subscription

### PR DESCRIPTION
**Pull request by paddynski-opencast**

### CODE REVIEW

File: app/converters/subscription/update/SubscriptionUpdateConverter.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+class SubscriptionUpdateConverter
+    extends BaseConverter[ServiceCreateRequests.Request, ConnectorUpdateRequests.Request, ConnectorResponses.Response, ServiceResponses.Response] {
+
+  override def convertServiceRequest(request: ServiceCreateRequests.Request): ConnectorUpdateRequests.Request =
+    ConnectorUpdateRequests.Request(id = request.id, name = request.name, contacts = request.contacts.map(convert))
+
+  override def convertSuccessfulConnectorResponse(response: ConnectorResponses.Response): ServiceResponses.Response =
+    ServiceResponses.Response("OK")
+
+  private def convert(contact: ServiceDataRequests.Contact): ConnectorDataRequests.Contact =
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/converters/subscription/update/SubscriptionUpdateConverter.scala:34]: Wrote `abstract class`.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,2641 @@
...
+import services.BaseService.Responses.Error
+import services.subscription.SubscriptionService
+import services.{BaseBackendConnectorSpec, BaseService}
+
+class SubscriptionUpdateServiceSpec extends BaseBackendConnectorSpec {
+
+  private val connectorPath: String = SubscriptionUpdateConnector.connectorPath
+  private lazy val service          = app.injector.instanceOf[SubscriptionUpdateService]
+
+  val id = "a7405c8d-06ee-46a3-b5a0-5d65176360ed"
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:31]: Search and replace? 

The harm - slight as it is - is that the test reader has to follow the reference.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,2641 @@
...
+          val response = await(service.call(request))
+
+          response shouldBe Right(SubscriptionService.Responses.Response("OK"))
+          verifyThatDownstreamApiWasCalled()
+        }
+      }
+
+      "fails, where the response body is" - {
+        "absent, with a status code of" - {
+          "service unavailable" in {
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:189]: There id s `SERVICE_UNAVAILABLE` in https://github.com/hmrc/dprs/blob/6317f6d2239924ac4c65e7b8e5e961f32bee2d5e/it/test/uk/gov/hmrc/dprs/subscription/UpdateSubscriptionSpec.scala

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,2641 @@
...
+                Seq(
+                  Error("eis-returned-service-unavailable")
+                )
+              )
+            )
+            verifyThatDownstreamApiWasCalled()
+          }
+        }
+
+        "valid, with a status code of" - {
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:264]: I copied all the scenarios in https://github.com/hmrc/dprs/blob/6317f6d2239924ac4c65e7b8e5e961f32bee2d5e/it/test/uk/gov/hmrc/dprs/subscription/UpdateSubscriptionSpec.scala

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,2641 @@
...
+                NOT_FOUND,
+                Seq(
+                  Error("eis-returned-not-found")
+                )
+              )
+            )
+            verifyThatDownstreamApiWasCalled()
+          }
+        }
+
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:399]: Done.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,2641 @@
...
+        "invalid, specifically:" - {
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:400]: I copied all the scenarios in https://github.com/hmrc/dprs/blob/6317f6d2239924ac4c65e7b8e5e961f32bee2d5e/it/test/uk/gov/hmrc/dprs/subscription/UpdateSubscriptionSpec.scala as specified on the ticket.

File: app/connectors/subscription/update/SubscriptionUpdateConnector.scala [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+class SubscriptionUpdateConnector @Inject() (frontendAppConfig: FrontendAppConfig, wsClient: WSClient)
+    extends BaseBackendConnector[Request, Response](frontendAppConfig, wsClient) {
+
+  override def connectorPath: String = SubscriptionUpdateConnector.connectorPath
+
+  override def call(request: Request)(implicit
+    executionContext: ExecutionContext
+  ): Future[Either[BaseConnector.Responses.Errors, Response]] =
+    post(fullUrl(request), request).map {
+      case Left(error) if error.status == NO_CONTENT => Right(Response("OK"))
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/connectors/subscription/update/SubscriptionUpdateConnector.scala:42]: Done.

**Comment by erasmos-hmrc-new** [comment]
@paddynski-opencast , could you update as advised?


SBT can be updated to 1.9.9 in [project/build.properties](https://github.com/hmrc/dprs-frontend/blob/ODPR-649/project/build.properties#L1)

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:376]: Perhaps I should have been more precise: I meant that these tests should reflect the various possible responses from the backend, in terms of status codes.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala]: Confirmed.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:248]: Perhaps I should have been more precise: I meant that these tests should reflect the various possible responses from the backend, in terms of status codes.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:376]: Perhaps I should have been more precise: I meant that these tests should reflect the various possible responses from the backend, in terms of status codes.

[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:376]: This is where the response (JSON) payload is invalid.

See `SubscriptionCreationServiceSpec`.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:248]: Perhaps I should have been more precise: I meant that these tests should reflect the various possible responses from the backend, in terms of status codes.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:177]: Yes, that's true: it is one of the status codes.

Just wondering why you went with that one, as opposed to Internal Server Error; though it does the job.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:31]: Unchanged.

File: it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala [hunk 8]
```diff
@@ -21,7 +21,7 @@ import connectors.BaseConnector
 import connectors.subscription.create.SubscriptionCreationConnector
 import play.api.http.Status.{BAD_REQUEST, CONFLICT, INTERNAL_SERVER_ERROR, OK, SERVICE_UNAVAILABLE}
 import services.BaseService.Responses.Error
-import services.subscription.create
+import services.subscription.{create, SubscriptionService}
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala:24]: Unchanged.

File: verify.sh [hunk 9]
```diff
@@ -1 +1 @@
-sbt "clean;scalafmtCheck;test:scalafmtCheck;it/scalafmtCheck;coverage;test;it/test;coverageReport"
\ No newline at end of file
+sbt "clean;scalafmtAll;test:scalafmtAll;it/scalafmtAll;coverage;test;it/test;coverageReport"
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply verify.sh:1]: Unchanged.

File: app/connectors/subscription/SubscriptionConnector.scala [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+          (JsPath \ "name").write[String] and
+          (JsPath \ "landline").writeNullable[String] and
+          (JsPath \ "mobile").writeNullable[String] and
+          (JsPath \ "emailAddress").write[String])(unlift(Organisation.unapply))
+    }
+  }
+
+  object Responses {
+
+    final case class Response(id: String)
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/SubscriptionConnector.scala:77]: Unchanged.

File: app/converters/subscription/update/SubscriptionUpdateConverter.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/update/SubscriptionUpdateConverter.scala]: Confirmed.

File: app/converters/subscription/create/SubscriptionCreationConverter.scala [hunk 11]
```diff
@@ -16,40 +16,36 @@
...
-    case ServiceRequests.IdType.SAFE => "SAFE"
-    case ServiceRequests.IdType.UTR  => "UTR"
+  private def convert(idType: ServiceCreateRequests.IdType): String = idType match {
+    case ServiceCreateRequests.IdType.NINO => "NINO"
+    case ServiceCreateRequests.IdType.SAFE => "SAFE"
+    case ServiceCreateRequests.IdType.UTR  => "UTR"
   }
 
-  private def convert(contact: ServiceRequests.Contact): SubscriptionCreationConnector.Requests.Contact =
+  private def convert(contact: ServiceDataRequests.Contact): ConnectorDataRequests.Contact =
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/converters/subscription/create/SubscriptionCreationConverter.scala]: Confirmed.

File: app/connectors/subscription/update/SubscriptionUpdateConnector.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/update/SubscriptionUpdateConnector.scala]: Confirmed.

File: app/connectors/BaseConnector.scala [hunk 12]
```diff
@@ -33,13 +33,13 @@ abstract class BaseConnector[REQUEST, RESPONSE](wsClient: WSClient) {
 
   /** We would have liked to use HttpClientV2, but when it encounters a 400 or 500 status code, the response body is inaccessible.
     */
-  def post(request: REQUEST)(implicit
+  def post(url: URL, request: REQUEST)(implicit
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/BaseConnector.scala:36]: Confirmed.

File: app/connectors/subscription/update/SubscriptionUpdateConnector.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/update/SubscriptionUpdateConnector.scala]: In my [proposal](https://github.com/hmrc/dprs-frontend/blob/ODPR-649-SEAN/app/connectors/subscription/update/SubscriptionUpdateConnector.scala#L39C53-L39C68), this returned an Option[Response].

How is this better?

Also, the response from the backend doesn't have the ID, so it shouldn't be here; in fact, it should return a None.

File: app/connectors/BaseConnector.scala [hunk 13]
```diff
@@ -71,6 +85,8 @@ object BaseConnector {
 
   object Exceptions {
     final class ResponseParsingException extends RuntimeException
+    final class ResponseExpectedException extends RuntimeException
```
Comment (erasmos-hmrc-new) [inline app/connectors/BaseConnector.scala:88]: We won't need these.

File: app/converters/subscription/SubscriptionConverter.scala [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+package converters.subscription
+
+import converters.BaseConverter
+import connectors.subscription.SubscriptionConnector.{Requests => ConnectorDataRequests, Responses => ConnectorResponses}
+import services.subscription.SubscriptionService.{Requests => ServiceDataRequests, Responses => ServiceResponses}
+
+abstract class SubscriptionConverter[SERVICE_REQUEST, CONNECTOR_REQUEST]
+    extends BaseConverter[SERVICE_REQUEST, CONNECTOR_REQUEST, ConnectorResponses.Response, ServiceResponses.Response] {
+
+  override def convertSuccessfulConnectorResponse(response: ConnectorResponses.Response): ServiceResponses.Response =
```
Comment (erasmos-hmrc-new) [inline app/converters/subscription/SubscriptionConverter.scala:26]: This isn't needed for the 'update' case, and fails to return the  ID for the 'create' case. We must have a gap in the tests?

File: test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+import java.util.UUID
+
+class SubscriptionUpdateConverterSpec extends BaseSpec {
+
+  private val converter = new SubscriptionUpdateConverter
+
+  "when updating a subscription, the converter returns the expected value, for a" - {
+    "service request" in {
+      val id1 = UUID.randomUUID().toString
```
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala]: Done.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:31]: Removed the `val id`.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:376]: Created a test where the response JSON is invalid.

File: it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala:24]: Optimised the imports.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:248]: Wrote a test that verifies each error code one at a time. It does not natter what the request is as we are testing that  the response error code is propagated from the connector to the service.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala]: Confirmed.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:240]: I don't understand the relevant of your comment about the request,  but agree with the intent of these scenarios.

Per my initial comment: I can see that you've added scenarios for both 'service unavailable' and 'conflict' but are still missing those for `forbidden` and `unauthorised`.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala]: Confirmed.

File: test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala]: Confirmed.

However, there is an indentation problem: "connector response" should be a sibling of "service request".

File: app/connectors/BaseConnector.scala (hunk 13 above)
Comment (erasmos-hmrc-new) [reply app/connectors/BaseConnector.scala:88]: Unchanged.

File: app/converters/subscription/SubscriptionConverter.scala (hunk 14 above)
Comment (erasmos-hmrc-new) [reply app/converters/subscription/SubscriptionConverter.scala]: Unchanged.

File: it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/create/SubscriptionCreationServiceSpec.scala]: Confirmed.

File: verify.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply verify.sh]: Confirmed.

File: app/connectors/subscription/SubscriptionConnector.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/SubscriptionConnector.scala]: Confirmed.

**Comment by erasmos-hmrc-new** [comment]
@paddynski-opencast , please use [this branch](https://github.com/hmrc/dprs-frontend/tree/ODPR-649-SEAN) I made, as a reference; it should assist you in resolving the above issues.

File: test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala]: Corrected.

File: app/connectors/subscription/update/SubscriptionUpdateConnector.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply app/connectors/subscription/update/SubscriptionUpdateConnector.scala]: Addressed.

File: app/converters/subscription/SubscriptionConverter.scala (hunk 14 above)
Comment (paddynski-opencast) [reply app/converters/subscription/SubscriptionConverter.scala]: Addressed.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (paddynski-opencast) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:240]: Added a test for the various status codes.

[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala:239]: In my guide branch, I had: "fails, with a status code of" /   "bad request" etc ... in this section, we're only interested in how various status codes are handled. The error responses were to be correctly formed, but in your version, they are not.

File: test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply test/converters/subscription/update/SubscriptionUpdateConverterSpec.scala]: Confirmed.

File: app/connectors/BaseConnector.scala (hunk 13 above)
Comment (erasmos-hmrc-new) [reply app/connectors/BaseConnector.scala]: Confirmed.

File: app/connectors/subscription/update/SubscriptionUpdateConnector.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply app/connectors/subscription/update/SubscriptionUpdateConnector.scala]: Confirmed.

**Comment by erasmos-hmrc-new** [comment]
I've added a comment, as agreed with @paddynski-opencast , to resolve the remaining issues.

File: it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (erasmos-hmrc-new) [reply it/test/services/subscription/update/SubscriptionUpdateServiceSpec.scala]: Resolved with my commit.

**Review by erasmos-hmrc-new — APPROVED** [review]

**Comment by erasmos-hmrc-new** [comment]
As I had added a commit, neither @paddynski-opencast or I are eligible to merge (quite rightly), so I've asked @Alex-H-94  to do so.

**Review by Alex-H-94 — APPROVED** [review]
As requested by Sean.

**Pull request merged**